core/vm/evmc.go

```diff
@@ -0,0 +1,280 @@
+	}
+	return common.Hash{}
+}
+
+func (host *HostContext) EmitLog(addr common.Address, topics []common.Hash, data []byte) {
+	env := host.env
+	env.StateDB.AddLog(&types.Log{
+		Address:     addr,
+		Topics:      topics,
+		Data:        data,
+		BlockNumber: env.BlockNumber.Uint64(),
+	})
+}
+
+func (host *HostContext) Call(kind evmc.CallKind,
+	destination common.Address, sender common.Address, value *big.Int, input []byte, gas int64, depth int,
+	static bool) (output []byte, gasLeft int64, createAddr common.Address, err error) {
+
+	env := host.env
+
+	gasU := uint64(gas)
+	var gasLeftU uint64
+
+	switch kind {
+	case evmc.Call:
+		if static {
+			output, gasLeftU, err = env.StaticCall(host.contract, destination, input, gasU)
+		} else {
+			output, gasLeftU, err = env.Call(host.contract, destination, input, gasU, value)
+		}
+	case evmc.DelegateCall:
+		output, gasLeftU, err = env.DelegateCall(host.contract, destination, input, gasU)
+	case evmc.CallCode:
+		output, gasLeftU, err = env.CallCode(host.contract, destination, input, gasU, value)
+	case evmc.Create:
+		var createOutput []byte
+		createOutput, createAddr, gasLeftU, err = env.Create(host.contract, input, gasU, value)
+		isHomestead := env.ChainConfig().IsHomestead(env.BlockNumber)
+		if !isHomestead && err == ErrCodeStoreOutOfGas {
+			err = nil
+		}
+		if err == errExecutionReverted {
+			// Assign return buffer from REVERT.
+			// TODO: Bad API design: return data buffer and the code is returned in the same place. In worst case
+			//       the code is returned also when there is not enough funds to deploy the code.
+			output = createOutput
+		}
+	}
+
+	// Map errors.
+	if err == errExecutionReverted {
+		err = evmc.Revert
+	} else if err != nil {
+		err = evmc.Failure
+	}
+
+	gasLeft = int64(gasLeftU)
+	return output, gasLeft, createAddr, err
+}
+
+func getRevision(env *EVM) evmc.Revision {
+	n := env.BlockNumber
+	if env.ChainConfig().IsByzantium(n) {
+		return evmc.Byzantium
+	}
+	if env.ChainConfig().IsEIP158(n) {
+		return evmc.SpuriousDragon
+	}
+	if env.ChainConfig().IsEIP150(n) {
+		return evmc.TangerineWhistle
+	}
+	if env.ChainConfig().IsHomestead(n) {
+		return evmc.Homestead
+	}
+	return evmc.Frontier
+}
+
+func (evm *EVMC) Run(contract *Contract, input []byte) (ret []byte, err error) {
+	evm.env.depth++
+	defer func() { evm.env.depth-- }()
+
+	// Don't bother with the execution if there's no code.
+	if len(contract.Code) == 0 {
+		return nil, nil
+	}
+
+	output, gasLeft, err := evm.instance.Execute(
+		&HostContext{evm.env, contract},
+		getRevision(evm.env),
+		contract.Address(),
+		contract.Caller(),
+		common.BigToHash(contract.value),
+		input,
+		crypto.Keccak256Hash(contract.Code),
+		int64(contract.Gas),
+		evm.env.depth-1,
+		evm.readOnly,
+		contract.Code)
+
+	contract.Gas = uint64(gasLeft)
+
+	if err == evmc.Revert {
+		err = errExecutionReverted
+	}
+
+	return output, err
+}
+
+func (evm *EVMC) CanRun([]byte) bool {
+	return true
```

Comment:
So it should be `false` and then never run? EVMC is not for ewasm only, I'm using it with C++ EVM interpreter. I believe if an user specifies the execution as `geth --vm myvm.so` he really wants to use `myvm.so`.

You should rethink this `CanRun()` feature.

Reply:
the feature has been thought through enough, thank you. If you have `geth --vm myvm.so` you are stuck with only one interpreter, whereas this supports a mixed chain with different formats.

Reply:
This will be handled with EVMC 6.

Reply:
then in EVMC-6, and provided the right command line switch, the return value here should be true. Until then, it should check if the format is wasm and, if not, return false.

Reply:
This would break geth + aleth-interpreter.

Reply:
Please be more specific. How would that break geth + aleth ?

Reply:
I need geth + aleth-interpreter for testing. Aleth is EVM and if we check for wasm bytecode here it will be skip and the execution fallback to Go Interpreter.

Reply:
Anyway, I'm not aiming to merge this PR now. We should definitely wait for EVMC 6 which should be soon.

Reply:
ok let's do that. In the mean time, here is the PR that implements all the command line switches: https://github.com/ethereum/go-ethereum/pull/17687